### PR TITLE
Resolve RDS connection issues, again

### PR DESCRIPTION
**Pull request by rymarczy**
This PR is a collection of changes that I made will attempting to resolve RDS connection issues.

The main RDS connection resolution is to limit the number of connections that are being created by our database engine object. This is mostly achieved by passing specific parameters to the sqlalchemy `create_engine` function.

Various logging updates are also included to aid in splunk debugging.

Asana Task: https://app.asana.com/0/1203185253920158/1203344379200929

### CODE REVIEW

File: performance_manager/lib/logging_utils.py [hunk 1]
```diff
@@ -18,11 +19,11 @@ def __init__(self, process_name: str, **metadata: MdValues) -> None:
...
         """
         self.parent = "performance_manager"
-
         self.process_name = process_name
-        self.uuid = uuid.uuid4()
-        self.start_time = time.time()
         self.metadata = metadata
+        self.uuid: Optional[uuid.UUID] = None
+        self.start_time = 0.0
+        self.process_id = 0
```
Comment (rymarczy) [inline performance_manager/lib/logging_utils.py:26]: move these initializations into `log_start`

File: performance_manager/lib/logging_utils.py [hunk 2]
```diff
@@ -31,10 +32,14 @@ def add_metadata(self, **metadata: MdValues) -> None:
 
     def log_start(self) -> None:
         """log the start of a proccess"""
+        self.uuid = uuid.uuid4()
+        self.start_time = time.monotonic()
+        self.process_id = os.getpid()
```
Comment (rymarczy) [inline performance_manager/lib/logging_utils.py:37]: add process id as standard for easier multiprocessing debugging

File: performance_manager/lib/postgres_utils.py [hunk 3]
```diff
@@ -30,15 +30,13 @@ def get_db_password() -> str:
...
-                Port=db_port,
-                DBUsername=db_user,
-                Region=db_region,
-            )
+        return client.generate_db_auth_token(
+            DBHostname=db_host,
+            Port=db_port,
+            DBUsername=db_user,
+            Region=db_region,
         )
```
Comment (rymarczy) [inline performance_manager/lib/postgres_utils.py:40]: not supposed to use `urlparse` here

File: performance_manager/lib/postgres_utils.py [hunk 4]
```diff
@@ -54,7 +52,10 @@ def postgres_event_update_db_password(
     update database passord on every new connection attempt
     this will refresh db auth token passwords
     """
+    process_logger = ProcessLogger("password_refresh")
+    process_logger.log_start()
     cparams["password"] = get_db_password()
+    process_logger.log_complete()
```
Comment (rymarczy) [inline performance_manager/lib/postgres_utils.py:58]: give us a log everytime connection password is requested

File: py_gtfs_rt_ingestion/ingest.py [hunk 5]
```diff
@@ -81,8 +82,7 @@ def validate_environment() -> None:
         )
 
 
-@schedule.repeat(schedule.every(5).minutes)
-def ingest() -> None:
+def ingest(metadata_queue: Queue) -> None:
```
Comment (rymarczy) [inline py_gtfs_rt_ingestion/ingest.py:85]: changed the way function is scheduled so we can pass `metadata_queue` parameter

File: py_gtfs_rt_ingestion/ingest.py [hunk 6]
```diff
@@ -98,29 +98,28 @@ def ingest() -> None:
...
 
 
 def main() -> None:
     """every second run jobs that are currently pending"""
-    ingest()
+    # start rds writer process
+    # this will create only one rds engine while app is running
+    metadata_queue, _rds_process = start_rds_writer_process()
+
+    schedule.every(5).minutes.do(ingest, metadata_queue=metadata_queue)
```
Comment (rymarczy) [inline py_gtfs_rt_ingestion/ingest.py:112]: create one `rds_writer_process` for life of application runtime, pass queue to `ingest`

File: py_gtfs_rt_ingestion/ingest.py [hunk 7]
```diff
@@ -98,29 +98,28 @@ def ingest() -> None:
...
+
     while True:
         schedule.run_pending()
         time.sleep(1)
 
+    # send shutdown signal to rds process and wait for finish
+    # not sure if we'll ever get here the way we're using @schedule
+    # metadata_queue.put(None)
+    # _rds_process.join()
+
```
Comment (rymarczy) [inline py_gtfs_rt_ingestion/ingest.py:122]: might need to un-comment this based on how app kill switch is implemented, this should gracefully shut down `rds_writer_process`

File: py_gtfs_rt_ingestion/lib/convert_gtfs_rt.py [hunk 8]
```diff
@@ -66,7 +66,9 @@ def __init__(self, config_type: ConfigType, metadata_queue: Queue) -> None:
 
     def convert(self) -> None:
         process_logger = ProcessLogger(
-            "gtfs_rt_convert", config_type=str(self.config_type)
+            "gtfs_rt_convert",
+            config_type=str(self.config_type),
+            file_count=len(self.files),
```
Comment (rymarczy) [inline py_gtfs_rt_ingestion/lib/convert_gtfs_rt.py:71]: log how many files are attempting to be converted

File: py_gtfs_rt_ingestion/lib/convert_gtfs_rt.py [hunk 9]
```diff
@@ -83,11 +85,14 @@ def process_files(self) -> Iterable[pyarrow.table]:
         iterate through all of the files to be converted, yielding a new table
         everytime the timestamps cross over an hour.
         """
-        table = self.detail.empty_table()
+        table = pyarrow.table(
+            self.detail.empty_table(),
+            schema=self.detail.export_schema,
+        )
```
Comment (rymarczy) [inline py_gtfs_rt_ingestion/lib/convert_gtfs_rt.py:91]: go back to old table coversion method, using all `dict` objects may have been causing run away memory use

File: py_gtfs_rt_ingestion/lib/convert_gtfs_rt.py [hunk 10]
```diff
@@ -124,36 +128,21 @@ def process_files(self) -> Iterable[pyarrow.table]:
...
-                    table = self.detail.empty_table()
+                    table = pyarrow.table(
+                        self.detail.empty_table(),
+                        schema=self.detail.export_schema,
+                    )
                     current_time = timestamp
                     file_count = 0
 
+                    process_logger.add_metadata(file_count=0)
                     process_logger.log_start()
```
Comment (rymarczy) [inline py_gtfs_rt_ingestion/lib/convert_gtfs_rt.py:143]: reset `file_count` in process logger

File: py_gtfs_rt_ingestion/lib/logging_utils.py [hunk 11]
```diff
@@ -18,11 +19,11 @@ def __init__(self, process_name: str, **metadata: MdValues) -> None:
         and uuid will be created for timing and unique identification
         """
         self.parent = "gtfs_ingestion"
-
         self.process_name = process_name
-        self.uuid = uuid.uuid4()
-        self.start_time = time.time()
         self.metadata = metadata
+        self.uuid: Optional[uuid.UUID] = None
```
Comment (rymarczy) [inline py_gtfs_rt_ingestion/lib/logging_utils.py:24]: same changes as in `performance_manager` logging utils

File: py_gtfs_rt_ingestion/lib/postgres_utils.py [hunk 12]
```diff
@@ -33,22 +32,21 @@ def get_db_password() -> str:
...
-                DBHostname=db_host,
-                Port=db_port,
-                DBUsername=db_user,
-                Region=db_region,
-            )
+        return client.generate_db_auth_token(
+            DBHostname=db_host,
+            Port=db_port,
+            DBUsername=db_user,
+            Region=db_region,
```
Comment (rymarczy) [inline py_gtfs_rt_ingestion/lib/postgres_utils.py:49]: same changes as other `postgres_utils`

File: py_gtfs_rt_ingestion/lib/s3_utils.py [hunk 13]
```diff
@@ -80,13 +81,6 @@ def _move_s3_object(filename: str, to_bucket: str) -> None:
     :param filename - expected as 's3://my_bucket/the/path/to/the/file.json
     :param to_bucket bucket name
     """
-    process_logger = ProcessLogger(
-        "move_s3_object",
-        filename=filename,
-        to_bucket=to_bucket,
-    )
-    process_logger.log_start()
-
```
Comment (rymarczy) [inline py_gtfs_rt_ingestion/lib/s3_utils.py:89]: remove individual object move logging, was spamming splunk output

function now returns `1` if move was success, `0` if failure, fail count given in `move_s3_objects` log

File: py_gtfs_rt_ingestion/lib/s3_utils.py [hunk 14]
```diff
@@ -164,11 +159,14 @@ def move_s3_objects(files: List[str], to_bucket: str) -> None:
...
     process_logger.log_start()
 
+    error_count = 0
     with Pool(pool_size, initializer=_init_process_session) as pool:
-        pool.starmap(
+        for pool_error_count in pool.starmap(
             _move_s3_object, [(filename, to_bucket) for filename in files]
-        )
+        ):
+            error_count += pool_error_count
```
Comment (rymarczy) [inline py_gtfs_rt_ingestion/lib/s3_utils.py:167]: collect count of move failures and log

File: performance_manager/lib/logging_utils.py (hunk 2 above)
Comment (mzappitello) [reply performance_manager/lib/logging_utils.py:37]: good idea.

File: performance_manager/lib/postgres_utils.py [hunk 15]
```diff
@@ -157,8 +159,15 @@ def get_local_engine(
...
-            pool_recycle=300,
             pool_pre_ping=True,
+            pool_use_lifo=True,
+            pool_size=5,
+            max_overflow=2,
+            connect_args={
+                "keepalives": 1,
+                "keepalives_idle": 60,
+                "keepalives_interval": 60,
+            },
```
[the remark this answers is not in the document]
Comment (mzappitello) [reply performance_manager/lib/postgres_utils.py:170]: 👍

File: py_gtfs_rt_ingestion/ingest.py (hunk 7 above)
Comment (mzappitello) [reply py_gtfs_rt_ingestion/ingest.py:122]: yeah. i'll look into that.

**Comment by mzappitello** [comment]
This all looks good to me. Plus its been working most of the day, so thats good too.

**Pull request merged**